lib/srv/discovery/database_watcher.go

```diff
@@ -138,16 +138,17 @@ func (s *Server) getCurrentDatabases() map[string]types.Database {
 func (s *Server) onDatabaseCreate(ctx context.Context, database types.Database) error {
 	s.Log.Debugf("Creating database %s.", database.GetName())
 	err := s.AccessPoint.CreateDatabase(ctx, database)
-	// If the database already exists but has an empty discovery group, update it.
-	if trace.IsAlreadyExists(err) && s.updatesEmptyDiscoveryGroup(
-		func() (types.ResourceWithLabels, error) {
+	// If the database already exists but has cloud origin and an empty
+	// discovery group, then update it.
+	if err != nil {
+		err := s.resolveCreateErr(err, types.OriginCloud, func() (types.ResourceWithLabels, error) {
```

Comment:
Discovery service should only be able to get databases with `OriginCloud`:
https://github.com/gravitational/teleport/blob/89be51492373e53715f8ca24f6cca90f4b6852c9/lib/authz/permissions.go#L1188-L1191

Are you able to grab the old database that has a different origin? Based on the above permission, Update should/supposed fail too as Discovery service shouldn't have permission to the previous object for Update call.

Reply:
I had db service running in the same process, maybe that's why it was allowed. Discovery needs to be able to Get any type of database. However I assumed that discovery service has this permission since it already calls GetDatabases for each reconciliation?

Reply:
My understanding is `GetDatabases` should only return the databases a user (or system user) has access to:
https://github.com/gravitational/teleport/blob/10932655f81c242cbbb56c76c29dae74486aff47/lib/auth/auth_with_roles.go#L5217-L5228

Related UT:
https://github.com/gravitational/teleport/blob/10932655f81c242cbbb56c76c29dae74486aff47/lib/auth/auth_with_roles_test.go#L2136-L2168

> I had db service running in the same process, maybe that's why it was allowed.

Which service has full RW on `types.KindDatabase`? Are you running Discovery with Auth? So per API call we can't tell which role it is calling with? that's tough

Reply:
We run discovery with auth service in cloud:
```
$ k exec -it deploy/teleport-auth -- tctl inventory status --connected                   (dev) 
Server ID                            Services       Version Upgrader 
------------------------------------ -------------- ------- -------- 
9d88ebf9-89ba-4076-99e1-f56f74ab235c Proxy          16.0.4  none     
3158eaa8-699f-4c5f-ab52-ec7e27fa8e75 Node           15.4.4  unit     
af09e479-3ecb-4295-bf1f-3745ded15f4c Auth,Discovery 16.0.4  none    
```

However, I could not reproduce the overwrite bug with my cloud tenant.

I am able to reproduce the bug if I run discovery service locally. I have several other services in there:
ssh, kube, db, app, discovery all in one config.

I think what's happening is that our "verb rules" check doesn't apply resource labels.
We just check if the identity is allowed to use VerbUpdate:
- https://github.com/gravitational/teleport/blob/f1dd572a2b55425ac1ecb7e3e601cb1289144bde/lib/auth/auth_with_roles.go#L6335-L6349

After that check passes, we check if they have "access" to the database, but the "access" check doesn't check the verb.

So basically, it's allowed to update the resource because `discovery_service` has the "update" verb allowed, and it's allowed to "access" the database because some other role like `db_service` has labels that match the database.

I'm not sure why it doesn't happen on cloud, but I suspect its due to the fact that auth doesn't use (nor need) a role with database labels in order to access databases.

I think for now at least, we have to assume there are some cases where discovery has permission to update the databases. It's a limitation of our RBAC.